iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import org.apache.thrift.TException;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SyncClientManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SyncClientManager.class);
+
+  private static final PipeConfig PIPE_CONFIG = PipeConfig.getInstance();
+
+  private final boolean useSSL;
+  private final String trustStore;
+  private final String trustStorePwd;
+  private final boolean useLeaderCache;
+
+  private final List<TEndPoint> endPoints;
+  private final Map<TEndPoint, Pair<IoTDBThriftSyncConnectorClient, Boolean>> endPoint2client =
+      new ConcurrentHashMap<>();
+
+  private final LeaderCacheManager leaderCacheManager = new LeaderCacheManager();
+
+  private long currentClientIndex = 0;
+
+  public SyncClientManager(
+      List<TEndPoint> endPoints,
+      boolean useSSL,
+      String trustStore,
+      String trustStorePwd,
+      boolean useLeaderCache) {
+    this.useSSL = useSSL;
+    this.trustStore = trustStore;
+    this.trustStorePwd = trustStorePwd;
+    this.useLeaderCache = useLeaderCache;
+    this.endPoints = endPoints;
+
+    for (TEndPoint endPoint : endPoints) {
+      endPoint2client.put(endPoint, new Pair<>(null, false));
+    }
+  }
+
+  public void initClients() throws IOException, TTransportException {
+    // init clients
+    for (TEndPoint endPoint : endPoints) {
+      if (Boolean.TRUE.equals(endPoint2client.get(endPoint).getRight())) {
+        continue;
+      }
+
+      initAndSetClient(endPoint);
+    }
+    // check whether any clients are available
+    for (TEndPoint nodeUrl : endPoint2client.keySet()) {
+      if (Boolean.TRUE.equals(endPoint2client.get(nodeUrl).getRight())) {
+        return;
+      }
+    }
+
+    throw new PipeConnectionException(
+        String.format("All target servers %s are not available.", endPoint2client.keySet()));
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getOneConnectedClientAndStatus() {
+    final int clientSize = endPoints.size();
+    // Round-robin, find the next alive client
+    for (int tryCount = 0; tryCount < clientSize; ++tryCount) {
+      final int clientIndex = (int) (currentClientIndex++ % clientSize);
+      Pair<IoTDBThriftSyncConnectorClient, Boolean> clientAndStatus =
+          endPoint2client.get(endPoints.get(clientIndex));
+      if (Boolean.TRUE.equals(clientAndStatus.getRight())) {
+        return clientAndStatus;
+      }
+    }
+    throw new PipeConnectionException(
+        "All clients are dead, please check the connection to the receiver.");
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeInsertNodeTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
```

Comment:
The name of the "endpoint" can be a little more explanatory, such as "leaderEndpoint".